### DISCUSSION TITLE
How can I use summarizeMomentGranularity

### DISCUSSION THREAD

**Discussion by arnaudcol** [pbeshai/tidy#27] [Q&A] [answered]
I noticed this summarize function in the documentation but I can't seem to call it.

**Comment by pbeshai** [comment, marked as answer]
Ah, good question. I'll update the docs to be more clear. In this case, since it depends on moment, you need to install a separate package. If you're using npm, you'd do this as follows:

```
npm install @tidyjs/tidy-moment moment
```

Then you could use it:

```js
import { summarizeMomentGranularity } from '@tidyjs/tidy-moment';
```

If you're using the unpkg CDN, you would import it: (note that the order of the script tags matters)

```html
<!-- tidy + dependencies -->
<script src="https://d3js.org/d3-array.v2.min.js"></script>
<script src="https://www.unpkg.com/@tidyjs/tidy/dist/umd/tidy.min.js"></script>

<!-- tidy-moment + dependencies -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.1/moment.min.js"></script>
<script src="https://www.unpkg.com/@tidyjs/tidy-moment/dist/umd/tidy-moment.min.js"></script>

<!-- code to use tidy and tidy moment -->
<script>
  const { tidy, sum } = Tidy;
  const { summarizeMomentGranularity } = TidyMoment;
  const data = [
    { str: "foo", date: moment.utc("2020-01-01"), value: 3 },
    { str: "foo", date: moment.utc("2020-01-03"), value: 1 },
    { str: "bar", date: moment.utc("2020-01-10"), value: 3 },
    { str: "bar", date: moment.utc("2020-01-21"), value: 1 },
    { str: "bar", date: moment.utc("2020-01-29"), value: 2 },
    { str: "bar", date: moment.utc("2020-02-01"), value: 5 }
  ];

  const results = tidy(
    data,
    summarizeMomentGranularity("weeks", { value: sum("value") })
  );
</script>
```

Let me know if you have any further difficulties!

**Comment by arnaudcol** [comment]
Hi,
I coped the code from:https://github.com/pbeshai/tidy/tree/main/packages/tidy-moment to test.  I am getting the following error in the console:
tidy-moment.min.js:1 Uncaught TypeError: Cannot read property 'first' of undefined
    at tidy-moment.min.js:1
    at z (tidy.min.js:1)
    at wiki-page.view?name=Trending2:343
    at handle (labkey.js?1882494251:245)
    at allDone (labkey.js?1882494251:656)
    at handle (labkey.js?1882494251:245)
    at Object.callbacksOnCache (labkey.js?1882494251:95)
    at HTMLScriptElement.cacheLoader (labkey.js?1882494251:723)

Any suggestions?

Thanks!

**Reply by pbeshai** [reply]
If i'm understanding you, it looks like you did not include the code for tidy itself? The full code to use it would be:

```html
<script src="https://d3js.org/d3-array.v2.min.js"></script>
<script src="https://www.unpkg.com/@tidyjs/tidy/dist/umd/tidy.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.1/moment.min.js"></script>
<script src="https://www.unpkg.com/@tidyjs/tidy-moment/dist/umd/tidy-moment.min.js"></script>
<script type="text/javascript">
  const { tidy, sum } = Tidy;
  const { summarizeMomentGranularity } = TidyMoment;
  const data = [
    { str: "foo", date: moment.utc("2020-01-01"), value: 3 },
    { str: "foo", date: moment.utc("2020-01-03"), value: 1 },
    { str: "bar", date: moment.utc("2020-01-10"), value: 3 },
    { str: "bar", date: moment.utc("2020-01-21"), value: 1 },
    { str: "bar", date: moment.utc("2020-01-29"), value: 2 },
    { str: "bar", date: moment.utc("2020-02-01"), value: 5 }
  ];

  const results = tidy(
    data,
    summarizeMomentGranularity("weeks", { value: sum("value") })
  );
  
</script>
```

**Reply by arnaudcol** [reply]
Thanks!  Looks like I was loading the scripts in the wrong order.

**Reply by pbeshai** [reply]
Ah yea, that's an easy mistake! glad it's working for you now.

**Reply by arnaudcol** [reply]
Fantastic project!  Combination of tidy.js with plotly.js makes some very nice visuals.

graphingData = tidy(LotList,
 leftJoin(OOSList, {by:'batch_no'}),
 filter((d) => d.harvest_date != null),
 mutate({
  date : (d) => moment(new Date(d.harvest_date)),
  oosnum: (d) => {if(d.oos){return 1}else{return 0}}
 }),
 summarizeMomentGranularity('w', { oosperiod: sum('oosnum') }),
 arrange(['harvest_date']),
 mutateWithSummary({
  rolling: roll(5, mean('oosperiod'), { partial: true }),
 }),
 select(['timestamp','oosperiod','rolling'])
)

![image](https://user-images.githubusercontent.com/15161284/112496621-226f2780-8d42-11eb-8441-e4b5cbb44403.png)

**Reply by pbeshai** [reply]
Nice work!! Using all the fancy features 😁
